rust/processor/src/db_writer.rs

```diff
@@ -0,0 +1,325 @@
+            _ => "DatabaseError::Other",
+        },
+        Error::NotFound => "NotFound",
+        Error::QueryBuilderError(_) => "QueryBuilderError",
+        Error::DeserializationError(_) => "DeserializationError",
+        Error::SerializationError(_) => "SerializationError",
+        Error::InvalidCString(_) => "InvalidCString",
+        Error::RollbackTransaction => "RollbackTransaction",
+        Error::AlreadyInTransaction => "AlreadyInTransaction",
+        Error::RollbackErrorOnCommit { .. } => "RollbackErrorOnCommit",
+        Error::NotInTransaction => "NotInTransaction",
+        Error::BrokenTransactionManager => "BrokenTransactionManager",
+        _ => "Other",
+    }
+}
+
+// This is a fake table for default fallback chunk size
+pub struct FakeTable {}
+impl field_count::FieldCount for FakeTable {
+    fn field_count() -> usize {
+        100
+    }
+}
+
+// A holder struct for processors db writing so we don't need to keep adding new params
+#[derive(Clone)]
+pub struct DbWriter {
+    pub processor_name: &'static str,
+    pub db_pool: PgDbPool,
+    pub query_sender: AsyncSender<QueryGenerator>,
+    pub per_table_chunk_sizes: AHashMap<String, usize>,
+    pub skip_tables: AHashSet<String>,
+}
+
+impl DbWriter {
+    pub fn new(
+        processor_name: &'static str,
+        db_pool: PgDbPool,
+        query_sender: AsyncSender<QueryGenerator>,
+        per_table_chunk_sizes: AHashMap<String, usize>,
+        skip_tables: AHashSet<String>,
+    ) -> Self {
+        Self {
+            processor_name,
+            db_pool,
+            query_sender,
+            per_table_chunk_sizes,
+            skip_tables,
+        }
+    }
+
+    pub fn chunk_size<Item: field_count::FieldCount>(&self, table_name: &str) -> usize {
+        get_config_table_chunk_size::<Item>(table_name, &self.per_table_chunk_sizes)
+    }
+
+    pub async fn send_in_chunks<'a, Query, Item>(
+        &self,
+        table_name: &'static str,
+        items_to_insert: Vec<Item>,
+        query_fn: fn(&'a [Item]) -> Query,
+    ) where
+        Item: field_count::FieldCount
+            + serde::Serialize
+            + for<'de> serde::Deserialize<'de>
+            + Send
+            + Sync
+            + Clone
+            + 'static,
+        Query:
+            QueryFragment<diesel::pg::Pg> + diesel::query_builder::QueryId + Send + Sync + 'static,
+    {
+        // TODO: ideally this kind of skip is done much earlier in the process
+        if self.skip_tables.contains(table_name) {
+            return;
+        }
+
+        DB_EXECUTOR_CHANNEL_CHUNK_INSERT_COUNT
+            .with_label_values(&[self.processor_name, table_name])
+            .inc();
+
+        let chunk_size = self.chunk_size::<Item>(table_name);
+        let chunks = get_chunks(items_to_insert.len(), chunk_size);
+        for (start_ind, end_ind) in chunks {
+            let items = items_to_insert[start_ind..end_ind].to_vec();
```

Comment:
is it a copy here?

Reply:
it is :-( I think there's likely a way to avoid it (and I didn't want `pop` or whatever) but I did not do it yet